framework/src/constraints/MortarConstraint.C

```diff
@@ -0,0 +1,251 @@
+#include "MooseVariable.h"
+#include "NearestNodeLocator.h"
+#include "PenetrationLocator.h"
+
+#include "libmesh/quadrature.h"
+
+template <>
+InputParameters
+validParams<MortarConstraint>()
+{
+  InputParameters params = validParams<Constraint>();
+  params.addRequiredParam<std::string>("interface", "The name of the interface.");
+  params.addRequiredParam<VariableName>("master_variable", "Variable on master surface");
+  params.addParam<VariableName>("slave_variable", "Variable on master surface");
+  return params;
+}
+
+MortarConstraint::MortarConstraint(const InputParameters & parameters)
+  : Constraint(parameters),
+    CoupleableMooseVariableDependencyIntermediateInterface(this, true),
+    _fe_problem(*getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
+    _dim(_mesh.dimension()),
+
+    _q_point(_assembly.qPoints()),
+    _qrule(_assembly.qRule()),
+    _JxW(_assembly.JxW()),
+    _coord(_assembly.coordTransformation()),
+    _current_elem(_assembly.elem()),
+
+    _master_var(_subproblem.getVariable(_tid, getParam<VariableName>("master_variable"))),
+    _slave_var(isParamValid("slave_variable")
+                   ? _subproblem.getVariable(_tid, getParam<VariableName>("slave_variable"))
+                   : _subproblem.getVariable(_tid, getParam<VariableName>("master_variable"))),
+    _lambda(_var.sln()),
+
+    _iface(*_mesh.getMortarInterfaceByName(getParam<std::string>("interface"))),
+    _master_penetration_locator(getMortarPenetrationLocator(
+        _iface._master, _iface._slave, Moose::Master, Order(_master_var.order()))),
+    _slave_penetration_locator(getMortarPenetrationLocator(
+        _iface._master, _iface._slave, Moose::Slave, Order(_slave_var.order()))),
+
+    _test_master(_master_var.phi()),
+    _grad_test_master(_master_var.gradPhi()),
+    _phi_master(_master_var.phi()),
+
+    _test_slave(_slave_var.phi()),
+    _grad_test_slave(_slave_var.gradPhi()),
+    _phi_slave(_slave_var.phi())
+{
+}
+
+void
+MortarConstraint::reinit()
+{
+  unsigned int nqp = _qrule->n_points();
+
+  _u_master.resize(nqp);
+  _grad_u_master.resize(nqp);
+  _phys_points_master.resize(nqp);
+  _u_slave.resize(nqp);
+  _grad_u_slave.resize(nqp);
+  _phys_points_slave.resize(nqp);
+  _test = _assembly.getFE(_var.feType(), _dim - 1)->get_phi(); // yes we need to do a copy here
+  _JxW_lm = _assembly.getFE(_var.feType(), _dim - 1)
+                ->get_JxW(); // another copy here to preserve the right JxW
+
+  for (_qp = 0; _qp < nqp; _qp++)
+  {
+    const Node * current_node = _mesh.getQuadratureNode(_current_elem, 0, _qp);
+
+    PenetrationInfo * master_pinfo =
+        _master_penetration_locator._penetration_info[current_node->id()];
+    PenetrationInfo * slave_pinfo =
+        _slave_penetration_locator._penetration_info[current_node->id()];
+
+    if (master_pinfo && slave_pinfo)
+    {
+      const Elem * master_side =
+          master_pinfo->_elem->build_side_ptr(master_pinfo->_side_num, true).release();
```

Comment:
Ugh.... Release and delete. Let's stop using that pattern. Just hold onto the smart point. If you hate it that much, just capture it in auto and you won't even see it. Yes, I realize that means you might have to us a `.get()`.

Reply:
You realize I didn't write any of this, right?  I just renamed some files here?

I don't really want to change this code at this point (not my code and I don't want to accidentally introduce any bugs)

Reply:
I am not worried about this, because the mortar stuff will be replaced by @jwpeterson's work soon. So no need to waste time here...